pkg/dns/treecache.go

```diff
@@ -0,0 +1,312 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dns
+
+import (
+	"bytes"
+	"crypto/md5"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"reflect"
+	"strings"
+	"sync"
+)
+
+const (
+	dataFile = "data.dat"
+	crcFile  = "data.crc"
+)
+
+type object interface{}
+
+type TreeCache struct {
+	ChildNodes map[string]*TreeCache
+	Entries    map[string]interface{}
+	m          *sync.RWMutex
+}
+
+func NewTreeCache() *TreeCache {
```

Comment:
function comment please, since this is a public symbol (though it probably doesn't need to be)